### PR TITLE
Cp ring update

### PR DESCRIPTION
**Pull request by stephen-hurd**

### CODE REVIEW

File: drivers/net/bnxt/bnxt_ring.c [hunk 1]
```diff
@@ -84,6 +85,9 @@ void bnxt_init_ring_grps(struct bnxt *bp)
  * tx bd ring - Only non-zero length if tx_ring_info is not NULL
  * rx bd ring - Only non-zero length if rx_ring_info is not NULL
  */
+//#define ROUNDUP(x) RTE_CACHE_LINE_ROUNDUP(x)
+#define ROUNDUP(size) \
+	(4096 * ((size + 4096 - 1) / 4096))
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_ring.c:90]: Why was this change required? Should this be page_size instead of 4096?

File: drivers/net/bnxt/bnxt_txr.c [hunk 2]
```diff
@@ -281,34 +282,50 @@ static void bnxt_tx_cmp(struct bnxt_tx_queue *txq, int nr_pkts)
...
 			if (CMP_TYPE(txcmp) == TX_CMPL_TYPE_TX_L2)
 				nb_tx_pkts++;
-			else
+			else {
 				RTE_LOG(DEBUG, PMD,
 						"Unhandled CMP type %02x\n",
 						CMP_TYPE(txcmp));
-			raw_cons = NEXT_RAW_CMP(raw_cons);
+
+				if (txcmp->flags_type & rte_cpu_to_le_16(1)) {
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_txr.c:309]: What does flags_type 0x1 signify? No-error Tx completion? May be a comment will help..

File: drivers/net/bnxt/bnxt_ethdev.c [hunk 3]
```diff
@@ -1332,17 +1332,13 @@ bnxt_dev_init(struct rte_eth_dev *eth_dev)
...
-	 * We can't forward commands before the VF driver calls drv_rgtr.
-	 * These are the ones that are may be used by drivers.
-	 */
-	ALLOW_FUNC(HWRM_VER_GET);
-	ALLOW_FUNC(HWRM_QUEUE_QPORTCFG);
-	ALLOW_FUNC(HWRM_FUNC_QCFG);
-	ALLOW_FUNC(HWRM_FUNC_QCAPS);
-	ALLOW_FUNC(HWRM_FUNC_DRV_RGTR);
+	/* Forward all requests if firmware is new enough */
+	if (1)
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_ethdev.c:1336]: So we will need a FW version check here?

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 4]
```diff
@@ -786,11 +789,11 @@ int bnxt_hwrm_stat_clear(struct bnxt *bp, struct bnxt_cp_ring_info *cpr)
 	struct hwrm_stat_ctx_clr_stats_input req = {.req_type = 0 };
 	struct hwrm_stat_ctx_clr_stats_output *resp = bp->hwrm_cmd_resp_addr;
 
-	HWRM_PREP(req, STAT_CTX_CLR_STATS, -1, resp);
-
 	if (cpr->hw_stats_ctx_id == (uint32_t)HWRM_NA_SIGNATURE)
 		return rc;
 
+	HWRM_PREP(req, STAT_CTX_CLR_STATS, -1, resp);
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_hwrm.c:795]: Want to make sure, this is not related to the CP ring processing. Just a reorg. Right?

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 5]
```diff
@@ -2014,8 +2020,8 @@ int bnxt_hwrm_allocate_vfs(struct bnxt *bp, int num_vfs)
 		rc = -ENOMEM;
 		goto error_free;
 	}
-	for (ui = 0; ui < req_buf_sz; ui += rte_eal_get_physmem_size()) {
-		rte_mem_lock_page(((char *)bp->pf.vf_req_buf) + ui);
+	for (sz = 0; sz < req_buf_sz; sz += getpagesize()) {
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_hwrm.c:2023]: This is a bug fix to use proper pagesize. right?
Just want to make sure. So that the patch can be split into CP ring processing, cleanup and other fix(es) while upstreaming.

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 6]
```diff
@@ -2130,6 +2136,23 @@ int bnxt_hwrm_tunnel_dst_port_free(struct bnxt *bp, uint16_t port,
...
+int bnxt_hwrm_func_cfg_def_cp(struct bnxt *bp)
+{
+	struct hwrm_func_cfg_output *resp = bp->hwrm_cmd_resp_addr;
+	struct hwrm_func_cfg_input req = {0};
+	int rc;
+
+	HWRM_PREP(req, FUNC_CFG, -1, resp);
+	req.fid = rte_cpu_to_le_16(0xffff);
+	req.flags = rte_cpu_to_le_32(bp->pf.func_cfg_flags);
+	req.enables = rte_cpu_to_le_32(HWRM_FUNC_CFG_INPUT_ENABLES_ASYNC_EVENT_CR);;
```
Comment (ajitkhaparde) [inline drivers/net/bnxt/bnxt_hwrm.c:2148]: Is this why the new firmware was required? Single semi-colon should be enough? ;-)

**Review by ajitkhaparde — APPROVED** [review]
Changes look good. But some comments/questions for my understanding.

File: drivers/net/bnxt/bnxt_ring.c (hunk 1 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_ring.c:90]: It's not clear at this time if the correct value is page size or 4096.  It is apparently most likely to be 4096.

File: drivers/net/bnxt/bnxt_txr.c (hunk 2 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_txr.c:309]: Per hsi_struct_defs_dpdk.h:

        /*
         * This field indicates the exact type of the completion. By
         * convention, the LSB identifies the length of the record in
         * 16B units. Even values indicate 16B records. Odd values
         * indicate 32B records.
         */

File: drivers/net/bnxt/bnxt_ethdev.c (hunk 3 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_ethdev.c:1336]: Yes, but the version number does not contain the information we need.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 4 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_hwrm.c:795]: Right, this is to prevent sequence number skipping spam in the chimp trace.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 5 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_hwrm.c:2023]: Yes, this is a separate bug fix.  I honestly don't know how I got this bit so wrong.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 6 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_hwrm.c:2148]: Heh, will fix.

The new firmware was required to allow VF forwarding when the VF driver is not registered.
